File: doubly-linked-lists/dll.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev  = None
        self.next  = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def show_vals(self):
        vals = []
        curr = self.head
        while curr:
            vals.append(curr.value)
            curr = curr.next
        return vals
    
    def push(self, value):
        newNode = Node(value)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        self.length += 1
        return self
    
    def pop(self):
        popped = None
        if self.length == 0:
            return popped
        elif self.length == 1:
            popped = self.tail
            self.head = None
            self.tail = None
        else:
            popped = self.tail
            self.tail = popped.prev
            self.tail.next = None
        self.length -= 1
        return self
    
    def unshift(self, value):
        newNode = Node(value)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head.prev = newNode
            self.head = newNode
        self.length += 1
        return self
    
    def shift(self):
        shifted = None
        if self.length == 0:
            return shifted
        elif self.length == 1:
            shifted = self.head
            self.head = None
            self.tail = None
        else:
            shifted = self.head
            self.head = shifted.next
            self.head.prev = None
            shifted.next = None
        self.length -= 1
        return shifted
    
    def get(self, position):
        if position < 0 | position >= self.length:
            return None
        elif position >= self.length / 2:
            i = self.length - 1
            target = self.tail
            while i != position:
                target = target.prev
                i -= 1
        else:
            i = 0
            target = self.head
            while i != position:
                target = target.next
                i += 1
        return target
    
    def set(self,value, position):
        target = self.get(position)
        if target:
            target.value = value
            return True
        return False
    
    def insert(self, value, position):
        if position < 0 | position > self.length:
            return None
        elif position == self.length:
            return self.push(value)
        elif position == 0:
            return self.unshift(value)
        newNode = Node(value)
        next= self.get(position)
        prev= next.prev
        newNode.prev = prev
        newNode.next = next
        prev.next = newNode
        next.prev = newNode
        self.length += 1
        return self
    
    def reverse(self):
        node = self.head
        self.head = self.tail
        self.tail = node
        prev = None
        next = None
        while node:
            next = node.next
            node.next = prev
            node.prev = next
            prev = node
            node = next
        return self


my_ll = DoublyLinkedList()
my_ll.push("A")
my_ll.push("B")
my_ll.push("C")
my_ll.reverse()

print(my_ll.show_vals())

File: doubly-linked-lists/dll.js
class Node {
    constructor(value) {
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

class DLL {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        let newNode = new Node(value);
        if (this.length === 0) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.prev = this.tail;
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    pop() {
        let popped = null;
        if (this.length === 0) return popped;
        if (this.head === this.tail) {
            popped = this.head;
            this.head = null;
            this.tail = null;
        }
        else {
            popped = this.tail;
            this.tail = popped.prev;
            this.tail.next = null;
        }
        this.length--;
        return popped;
    }

    unshift(value) {
        let newNode = new Node(value);
        if (this.length === 0) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
        this.length++;
        return this;
    }

    shift() {
        let shifted = null;
        if (this.length === 0) return shifted;
        if (this.head === this.tail) {
            shifted = this.head;
            this.head = null;
            this.tail = null;
        }
        else {
            shifted = this.head;
            this.head = shifted.next;
            this.head.prev = null;
        }
        this.length--;
        return shifted;
    }

    get(position) {
        let target, i;
        if (position < 0 || position >= this.length) return undefined;
        if (position > (this.length / 2)) {
            target = this.tail;
            i = this.length - 1;
            while (i != position) {
                target = target.prev;
                i--;
            }
        }
        else {
            target = this.head;
            i = 0;
            while (i != position) {
                target = target.next;
                i++;
            }
        }
        return target;
    }

    set(value, position) {
        let target = this.get(position);
        if (target) {
            target.value = value;
            return true;
        }
        return false;
    }

    insert(value, position) {
        if (position < 0 || position > this.length) return false;
        if (position === 0) {
            this.unshift(value);
            return true;
        }
        if (position === this.length) {
            this.push(value);
            return true;
        }
        let node = new Node(value);
        let next = this.get(position);
        let prev = next.prev;
        node.next = next;
        node.prev = prev;
        next.prev = node;
        prev.next = node;
        this.length++;
        return true;
    }

    remove(position) {
        if (position < 0 || position > this.length - 1) return undefined;
        if (position === 0) return this.shift();
        if (position === this.length - 1) return this.pop();
        let target = this.get(position);
        let prev = target.prev;
        let next = target.next;
        prev.next = next;
        next.prev = prev;
        this.length--;
        return target;
    }

    reverse() {
        let node = this.head;
        this.head = this.tail;
        this.tail = node;
        let prev = null;
        let next = null;
        for (let i = 0; i < this.length; i++) {
            next = node.next;
            node.next = prev;
            node.prev = next;
            prev = node;
            node = next;
        }
        return this;
    }
}

module.exports = DLL
